lua/metals/test_explorer.lua

```diff
@@ -0,0 +1,178 @@
+  location = {
+    range = {
+      start = { line = 0, character = 0 },
+      ["end"] = { line = 0, character = 0 },
+    },
+    uri = "",
+  },
+  targetName = "",
+  testCases = nil,
+}
+
+function TestSuite:new(suite, targetName)
+  suite = suite or {}
+  suite.targetName = targetName
+  setmetatable(suite, self)
+  self.__index = self
+  return suite
+end
+
+function TestSuite:add_test_cases(testCases)
+  self.testCases = testCases
+end
+
+local function handle_remove_suite(targetName, event)
+  state[targetName][event.fullyQualifiedClassName] = nil
+end
+
+local function handle_add_suite(targetName, event)
+  state[targetName][event.fullyQualifiedClassName] = TestSuite:new(event, targetName)
+end
+
+local function handle_update_suite_location(targetName, event)
+  state[targetName][event.fullyQualifiedClassName].location = event.location
+end
+
+local function handle_add_test_cases(targetName, event)
+  state[targetName][event.fullyQualifiedClassName]:add_test_cases(event.testCases)
+end
+
+local function handle(targetName, event)
+  if event.kind == "removeSuite" then
+    handle_remove_suite(targetName, event)
+  elseif event.kind == "addSuite" then
+    handle_add_suite(targetName, event)
+  elseif event.kind == "updateSuiteLocation" then
+    handle_update_suite_location(targetName, event)
+  elseif event.kind == "addTestCases" then
+    handle_add_test_cases(targetName, event)
+  else
+    log.error("Unknown event in test explorer", targetName, event)
+  end
+end
+
+local function get_test_suites(test_suite_uri)
+  local test_suites = {}
+  for _, buildTargetState in pairs(state) do
+    for _, test_suite in pairs(buildTargetState) do
+      if test_suite_uri == nil or test_suite.location.uri == test_suite_uri then
+        table.insert(test_suites, test_suite)
+      end
+    end
+  end
+  return test_suites
+end
+
+local function get_test_cases(test_suite)
+  if test_suite.testCases == nil then
+    return {}
+  end
+  local test_cases = {}
+  for _, test_case in pairs(test_suite.testCases) do
+    table.insert(test_cases, test_case)
+  end
+  return test_cases
+end
+
+local function dap_run_test(root, test_suite, test_case)
+  local dap_ok, dap = pcall(require, "dap")
+  if not dap_ok then
+    return log.error_and_show(messages.run_test_without_nvim_dap)
+  end
+  local tests_to_run = {}
+  if test_case ~= nil then
+    tests_to_run = { test_case.name }
+  end
+  local arguments = {
+    type = "scala",
+    request = "launch",
+    name = "Run Test",
+    -- ScalaTestSuitesDebugRequest
+    metals = {
+      target = { uri = "file:" .. root .. "/?id=" .. test_suite.targetName },
+      requestData = {
+        suites = {
+          {
+            className = test_suite.fullyQualifiedClassName,
+            tests = tests_to_run,
+          },
+        },
+        jvmOptions = {},
+        environmentVariables = {},
+      },
+    },
+  }
+  dap.run(arguments)
+end
+
+M.dap_select_test_suite = function(root)
```

Comment:
So for these I don't think we even need root. Because in `dap_run_test` we end up using that to put together the `uri` when actually we don't want to do that as it doesn't seem to work. Instead when the `executeClientCommand` comes in with the `metals-update-test-explorer` there is a `targetUri` that is sent with each target. Instead of using `root` here we need to use the `targetUri` instead.